traits.go

```diff
@@ -13,6 +13,72 @@ type Trait interface {
 	ValidateTrait() error
 }
 
+// OpenCloseTrait provides an implementation of the Smart Home OpenClose Trait Schema from Google Smart Home Actions
+type OpenCloseTrait struct {
+	DiscreteOnlyOpenClose bool
+	OpenDirection []string
+	QueryOnlyOpenClose bool
+	OnExecuteChange  OpenCloseCommand
+	OnDirectionStateHandler   func(Context) (string, proto.ErrorCode)
+	OnPercentStateHandler func(Context) (float64, proto.ErrorCode)
+}
+
+// ValidateTrait checks if all required attributes and handlers are created/set
+func (t OpenCloseTrait) ValidateTrait() error {
+	if t.OnExecuteChange == nil {
+		return errors.New("OnExecuteChange cannot be nil")
+	}
+	if t.OnDirectionStateHandler == nil || t.OnPercentStateHandler == nil {
+		return errors.New("Both OnStateHandlers must be filled cannot be nil")
+	}
+
+	return nil
+}
+
+// TraitName returns the string how google defines this traits name
+func (t OpenCloseTrait) TraitName() string {
+	return "action.devices.traits.OpenClose"
+}
+
+// TraitStates parses the diffrent state attributes and calls the corresponding handlers
+func (t OpenCloseTrait) TraitStates(ctx Context) []State {
+	var openPercentState State
+	openPercentState.Name = "openPercent"
+	openPercentState.Value, openPercentState.Error = t.OnPercentStateHandler(ctx)
+
+	// optional 
+	var openDirectionState State
+	openDirectionState.Name = "openDirection"
+	openDirectionState.Value, openDirectionState.Error = t.OnDirectionStateHandler(ctx) 
+
+	// check status handler
+	return []State{openPercentState, openDirectionState}
```

Comment:
I don't think this is correct, docs say you return a state called ```openState``` that is an array of objects, one for each ```openState```
https://developers.google.com/assistant/smarthome/traits/openclose#response

Reply:
you're right, fixed that